blockmanager.go

```diff
@@ -2743,29 +2763,42 @@ func areHeadersConnected(headers []*wire.BlockHeader) bool {
 	return true
 }
 
-// checkHeaderSanity checks the PoW, and timestamp of a block header.
+// checkHeaderSanity performs contextual and context-less checks on the passed
+// wire.BlockHeader. This function calls blockchain.CheckBlockHeaderContext for
+// the contextual check and blockchain.CheckBlockHeaderSanity for context-less
+// checks.
 func (b *blockManager) checkHeaderSanity(blockHeader *wire.BlockHeader,
-	maxTimestamp time.Time, reorgAttempt bool) error {
+	reorgAttempt bool, prevNodeHeight int32,
+	prevNodeHeader *wire.BlockHeader) error {
 
-	diff, err := b.calcNextRequiredDifficulty(
-		blockHeader.Timestamp, reorgAttempt)
-	if err != nil {
-		return err
+	// Create the lightHeaderCtx for the blockHeader's parent.
+	hList := b.headerList
+	if reorgAttempt {
+		hList = b.reorgList
 	}
-	stubBlock := btcutil.NewBlock(&wire.MsgBlock{
-		Header: *blockHeader,
-	})
-	err = blockchain.CheckProofOfWork(stubBlock,
-		blockchain.CompactToBig(diff))
+
+	parentHeaderCtx := newLightHeaderCtx(
+		prevNodeHeight, prevNodeHeader, b.cfg.BlockHeaders, hList,
+	)
+
+	// Create a lightChainCtx as well.
+	chainCtx := newLightChainCtx(
+		&b.cfg.ChainParams, b.blocksPerRetarget, b.minRetargetTimespan,
+		b.maxRetargetTimespan,
+	)
+
+	var emptyFlags blockchain.BehaviorFlags
+	err := blockchain.CheckBlockHeaderContext(
+		blockHeader, parentHeaderCtx, emptyFlags, chainCtx, true,
```

Comment:
Don't we also want to continue to verify the checkpoints here as well? Then we can remove the existing code to verify the checkpoints.

Reply:
I think I found it more complicated to also verify the checkpoints in btcd & would've rather had it in a different PR

Reply:
Yeah, we would need to implement `lightChainCtx.VerifyCheckpoint` and `lightChainCtx.FindPreviousCheckpoint`. I can do that, but it'll delay this PR since I'd also want to add unit tests around that.